Fix for phantom moves

Epiphero opened the pull request:
Sometimes (this seems to be a problem with webkit/Safari and magic mice, but has been observed on other platforms as well), a mousemove event is generated before the mouse actually moves.  This prevents a click gesture from being recognized upon a subsequent mouseup event.  The upshot is that clicking can be very difficult with most clicks just doing nothing.

This small patch checks that the mouse has indeed moved before setting the moved flag for the gesture or performing a transform.

mbostock commented:
See #65 and d3/d3-drag#28 for context. This is an improvement, but it only handles the case of ignoring strictly zero-movement mousemove events.

The problem is how we handle small but non-zero-movement mousemove events. I think the desired behavior is that small non-zero-movement mousemove events should still trigger a zoom event, but they should not trigger click suppression following the gesture (*i.e.*, *gesture*.moved should remain false until the mouse has moved a certain total distance since the start of the *gesture*, even though we may dispatch a *zoom* event).

Also, if and when a distance threshold is applied, that distance threshold must operate in screen space (*event*.clientX and *event*.clientY) not user space (d3.mouse), since user space can have an arbitrary transform associated with it. You need to do something like this (omitting the other parts of handling the gesture):

```js
function mousedowned() {
  var g = …,
        x0 = event.clientX,
        y0 = event.clientY,
        d2 = 0; // squared distance travelled during this gesture

  function mousemoved() {
    if (!g.moved) {
      var x1 = event.clientX,
          y1 = event.clientY,
          dx = x1 - x0,
          dy = y1 - y0;
      g.moved = (d2 += dx * dx + dy * dy) >= threshold2;
    }
  }
}
```

Where *threshold2* is the squared distance threshold to trigger click suppression. Probably something like 4.

I’ve also been thinking about whether this intersects with touch handling, but I don’t think it does, since touch handling doesn’t have the same click suppression logic.

Also, whatever strategy we apply to d3-zoom we need to apply the same strategy to d3-drag for consistency.

Epiphero commented:
Thanks for the context links, I should have spend a little more time digging around!

I've incorporated your suggestions:

- The `zoom.clickDistance()` accessor has been added to make the threshold configurable.  I've left the default at 0 to keep the behavior as close as possible to the current semantics while fixing the click problem on buggy zero-movement browsers.
- Changed to screen space from user space.
- Departing from your sample code, I don't compute the path distance, just the distance between the down and up.

If you think this is in line with your design goals, I'm happy to apply the same to d3-drag.

leon commented:
I'm having great troubles with this same problem, would it be possible to merge this?

mbostock commented:
Looks good. Thank you!
